crates/proto/src/xfer/dnssec_dns_handle.rs

```diff
@@ -291,7 +333,7 @@ where
 
     // collect all the rrsets to verify
     // TODO: is there a way to get rid of this clone() safely?
-    for (name, record_type) in rrset_types {
+    for (name, record_type) in rrset_types.clone() {
```

Comment:
So a bunch of this is a pre-existing issue, but IMO before adding more complexity here I think we should do a better job of organizing this code and/or commenting on why the complexity here is needed. In particular, do we really need four loops over the records and a bunch of clones?

- `for rrset in records.iter()` which builds `rrset_types`, `rrset_proofs`
- `for (name, record_type)` in rrset_types.clone()` with an inner loop for `rrs_to_verify` and an inner loop to build `rrsigs` which is apparently only used for logging?
- `for record in &mut records` to set the proofs of all the records
- `for record in &mut records` to update TTLs

Reply:
That's a valid point, personally I would suggest to refactor the function `verify_rrsets` in a separate PR.

After spending a bit of time to figure out how to represent the given data in a better way or how to eliminate loops / HashMaps, I had to stop for now. For one holding references over the `records` variable directly causes issues with the borrow checker, the `HashMap`s hold the relevant information (cloned or copied) over `records` and avoiding borrow checker issues. The `records` variable also holds relational information between DNSKEY, RRSIG and its covering RRset within its `Vec` which is currently not well represented or even obvious. The code block also does not know which RRSIG covers the given RRset, this information is determined inside [`verify_default_rrset`](https://github.com/hickory-dns/hickory-dns/blob/main/crates/proto/src/xfer/dnssec_dns_handle.rs#L584), but unavailable to the calling function. I'm quite sure this part of the code can be refactored to be cleaner and in a way that makes it more transparent what its purpose is. Imho a refactoring here will most likely touch other `verify_*` methods too.

Reply:
Doing that in a separate PR is fine with me, the question is if we review/merge that PR before this one or after. As such, this feels like a black hole of complexity and IMO refactoring it before making further changes would be worth it.

Reply:
There was good feedback in yesterday's discussion, the authenticated `ttl` value is stored alongside the `proof`, which eliminated the last `for` loop.

Reply:
@djc do you think the current logic is fine or would you rather have this block refactored further?

Reply:
I think this is good for now, thanks!